### PR TITLE
feat: similar_movies_search_implemented

### LINKED ISSUE #39: function Similar movies
D.9 

Design and implement a function that, given a movie as input, will provide a list of other movies that are similar

### Suggestions
- Check similarity between plots
- Genres
- Matching titles?
- Director
- Actors
- Country of origin

### PR DESCRIPTION
**Pull request by MarLaslo**
D9 
Created get_similar_movies(media_id)

Function checks for
-  for the movies from same country 
- different movies of same actors 
- different movies of same crew
- different movies with similar words in title

All of the checks have their own functions impleneted

>>get_count_of_movies_with_same_actors(media_id)
find_movies_from_the_same_country(country_id)
find_movie_similar_titles(title,min_lenght)
find_movies_with_similar_crew(media_id)

All functions return media_id and count of how many times has media appeared in a search result. 

`get_count_of_movies_with_same_actors(media_id)`
Gets the list of actors from the movie. Then it finds all the other movies the actors were part of and counts the overlapping media.

Same goes for 
`find_movies_with_similar_crew(media_id)`
But with the crew members.

`find_movies_from_the_same_country(country_id)`
This just gets all the movies from the same country and assignes 1 to each media_id from the country

`find_movie_similar_titles(title,min_lenght)`
This one takes string title. Splits it into words and searches for substring in other titles for each word in the title. Then just counts together the movies with same ids.

It also takes second parameter:
min_lenght - defines the minimum length the word in the title must have to be counted into the search,
it is set to 3, so it excludes words like "the, and, an. who..." and etc

THE FINAL FUNCTION IS a.k.a The Final Boss a.k.a the dominator a.k.a the one who possesses the holy grail a.k.a the one who can lick their elbow

Or just:
`get_similar_movies(media_id)`
This function calls all the other functions and returns a table which is list of media_id and sum of counts from other functions.

basically the higher the SUM the bigger chance that the movie will be similar

probably the function similar title will need the change after Sofus will finish the title separation task

### CODE REVIEW

File: src/D_functions_and_procedures.sql [hunk 1]
```diff
@@ -362,4 +362,192 @@ $$ LANGUAGE plpgsql;
...
+		SELECT DISTINCT cm.media_id, 1
+		FROM cast_member cm
+		WHERE cm.person_id = actor
+		GROUP BY cm.media_id;
+	END LOOP;
+	
+	RETURN query
+	SELECT s.media_id, count(s.media_id)::INTEGER
+	FROM similar_cast s
+	GROUP BY s.media_id;
```
Comment (idax6797) [inline src/D_functions_and_procedures.sql]: I maybe think it whould be nice to have it ordered by the count in DESC, when I guess you could see the most relevant media.

And also this is needs to be applied on all.

File: src/D_functions_and_procedures.sql [hunk 2]
```diff
@@ -362,4 +362,192 @@ $$ LANGUAGE plpgsql;
...
+	GROUP BY com_res.media_id
+	ORDER BY total_count DESC;
+END;
+$$ LANGUAGE plpgsql;
+
+-------------------------------------------------------
+---------get_count_of_movies_with_same_actors----------
+	
+CREATE OR REPLACE FUNCTION get_count_of_movies_with_same_actors(input_media_id INTEGER)
+RETURNS TABLE(
```
Comment (idax6797) [inline src/D_functions_and_procedures.sql:796]: Could you maybe also return the title of the media, this is for now done by joining the release table and finding the title that has the type original?


This is regarding all the functions

File: src/D_functions_and_procedures.sql [hunk 3]
```diff
@@ -362,4 +362,192 @@ $$ LANGUAGE plpgsql;
...
 -- D12 TEST
 SELECT * FROM best_match_titles(ARRAY['apple', 'mads', 'mikkelsen']);
-SELECT title FROM release WHERE media_id=47460;
\ No newline at end of file
+SELECT title FROM release WHERE media_id=47460;
+
+
+
+-------------------D9 SIMILAR MOVIES SEARCH-------------------------
+--------------------------------------------------------------------
```
Comment (idax6797) [inline src/D_functions_and_procedures.sql]: Maybe we need to find a common way to do this function separation so it looks alike in the file.

File: src/D_functions_and_procedures.sql [hunk 4]
```diff
@@ -362,4 +362,192 @@ $$ LANGUAGE plpgsql;
...
+	END LOOP;
+	
+	RETURN query
+	SELECT s.media_id, count(s.media_id)::INTEGER
+	FROM similar_crew s
+	GROUP BY s.media_id;
+END;
+$$ LANGUAGE plpgsql;
+
+--------------------------------------------------
```
Comment (idax6797) [inline src/D_functions_and_procedures.sql]: Could be nice to have under each function, then I at least understand what is what  - but nice work.

**Review by idax6797 — CHANGES_REQUESTED** [review]

File: src/D_functions_and_procedures.sql (hunk 2 above)
Comment (MarLaslo) [reply src/D_functions_and_procedures.sql:796]: done

File: src/D_functions_and_procedures.sql (hunk 3 above)
Comment (MarLaslo) [reply src/D_functions_and_procedures.sql]: I can make it in the final script

File: src/D_functions_and_procedures.sql [hunk 5]
```diff
@@ -368,39 +368,55 @@ SELECT title FROM release WHERE media_id=47460;
...
 	) AS com_res
 	GROUP BY com_res.media_id
 	ORDER BY total_count DESC;
+	
+	RETURN query
+	SELECT DISTINCT sr.media_id, sr.total_count, t.name
+	FROM search_result sr
+	JOIN title t USING (media_id)
+	JOIN title_title_type ttt USING (title_id)
+	WHERE ttt.title_type_id = 8
```
Comment (sofushn) [inline src/D_functions_and_procedures.sql:787]: I dont think you should hardcode id values like this, especially in a function

File: src/D_functions_and_procedures.sql [hunk 6]
```diff
@@ -428,7 +444,7 @@ BEGIN
 	FOREACH actor IN ARRAY COALESCE(actors, '{}')
 	LOOP
 		INSERT INTO similar_cast(media_id, "count")
-		SELECT DISTINCT cm.media_id, 1
+		SELECT cm.media_id, 1
```
Comment (sofushn) [inline src/D_functions_and_procedures.sql:816]: what is the point of selecting `1` here?

**Review by idax6797 — CHANGES_REQUESTED** [review]
Would order the function also so it is after D8 and not D12, that would be great :)

File: src/D_functions_and_procedures.sql [hunk 7]
```diff
@@ -368,39 +368,55 @@ SELECT title FROM release WHERE media_id=47460;
...
-	SELECT DISTINCT m.media_id, r.title, mpc.country_id INTO input_media
+	SELECT DISTINCT m.media_id, t.name, mpc.country_id INTO input_media
 	FROM media m
-	JOIN release r USING (media_id)
+	JOIN title t USING (media_id)
 	LEFT JOIN media_production_country mpc USING(media_id)
 	WHERE m.media_id = input_media_id;
 	
-	RETURN query
+	CREATE TEMPORARY TABLE search_result (
```
Comment (sofushn) [inline src/D_functions_and_procedures.sql:760]: You dont have to change it, but I think it would have been better to use the with statement here, I think that temporary tables only really should be used if you perform some kind of very expensive query. 

Also I think that the way its implemented here can create some problems if for example call the search function multiple times, then you would end up with a lot of temporary tables in the database which is not really scalable

**Review by sofushn — APPROVED** [review]
There are probably a couple of things that I would change if we had some more time but I think we should just get this merged and make sure that everything works so we are ready for hand in

File: src/D_functions_and_procedures.sql [hunk 8]
```diff
@@ -362,4 +606,4 @@ $$ LANGUAGE plpgsql;
 
 -- D12 TEST
 SELECT * FROM best_match_titles(ARRAY['apple', 'mads', 'mikkelsen']);
-SELECT title FROM release WHERE media_id=47460;
\ No newline at end of file
+SELECT title FROM release WHERE media_id=47460;
```
Comment (idax6797) [inline src/D_functions_and_procedures.sql]: This should also be removed, old stuff

File: src/D_functions_and_procedures.sql (hunk 5 above)
Comment (MarLaslo) [reply src/D_functions_and_procedures.sql:787]: so you want it to say 'original'?

File: src/D_functions_and_procedures.sql (hunk 6 above)
Comment (MarLaslo) [reply src/D_functions_and_procedures.sql:816]: missed that. this does not need to be there

File: src/D_functions_and_procedures.sql (hunk 7 above)
Comment (MarLaslo) [reply src/D_functions_and_procedures.sql:760]: are not they removed immediately after the function is done? got to read into it better.

**Review by idax6797 — APPROVED** [review]
